nengo_loihi/splitter.py

```diff
@@ -154,8 +154,9 @@ def place_ensembles(networks):
         # User-specified config takes precedence
         if config[ens].on_chip is not None:
             networks.move(ens, "chip" if config[ens].on_chip else "host")
-        # Direct mode ensembles must be off chip
-        elif isinstance(ens.neuron_type, nengo.Direct):
+        # Unsupported neuron types must be off-chip
+        elif not isinstance(
+                ens.neuron_type, (nengo.LIF, nengo.SpikingRectifiedLinear)):
```

Comment:
This is an interesting problem. On the one hand, I like that someone's model will just work. On the other hand, I don't like that an ensemble with a different neuron type will be silently moved off-chip. I feel like this quiet splitting is what's led to problems in the past (with us thinking more of the keyword spotter was on chip than actually was), and will likely confuse users as well (if I pass my model to `nengo_loihi`, I kind of assume it's going to run it almost exclusively on Loihi).

Also, it runs into the same problem as I pointed out above, where people can't add hooks for other neuron types. The way things are currently set up, this actually might not be as big of a problem as I was thinking because a) people can subclass off `LIF` or `SpikingRectifiedLinear` and b) there aren't really any single-compartment models that Loihi supports other than these two. However, as we open the door to multi-compartment, there may be either multi-compartment models that someone may want to add, or a way of doing a more complicated single-compartment model using multiple Loihi compartments.

Reply:
I agree that silently simulating these neurons off-chip will probably lead to confusion.  But it would also be nice to support, so that if someone really does intend to build a mixed model they can.  Maybe we could make this an explicit opt-in somehow? Something like
```
net.config[nengo.Izhikevich].on_chip = False
```
(building on our existing `on_chip` logic for ensembles).

Reply:
I guess we could have a registry of known neuron types that is based on the builder? That is I think quick enough that we should do it in this PR ... the problem of things not happening on chip that you think are happening on chip is a tough one, like should we raise a warning for every ensemble that won't be put on chip? Right now there's an exception if all ensembles are off-chip, which will hopefully be sufficient for the majority of cases where people think Loihi does something that it doesn't. But I think that particular issue should be deferred to another PR so as to not block the release, I'll implement the neuron type registry though so people can add other builders.

Reply:
Do we need this change for the nengo tests to work? Or is it an option to keep going as we were, with Ensembles that have a neuron type that is not supported on Loihi just raising an error? The option to put individual ensembles off-chip is of course there, so if someone has an ensemble with a different neuron type they can manually specify `on_chip = False`. It also wouldn't be hard for a user to loop over all ensembles in a model and set `on_chip = False` to any that have a specific neuron type, or don't have `LIF` or `SpikingRectifiedLinear`.

Reply:
It's not strictly required, the question is how you interpret a particular test failure. With this change, these failing tests fall in the "there are no ensembles being simulated on-chip" category, which is pretty clearly something that we are not planning to fix in Nengo Loihi. Before, the test failed with "hey there's no builder for this neuron type," which looks like something that we just haven't implemented yet.

One alternative would be to add builders for these neuron types that explicitly say "this neuron type cannot be simulated on Loihi. Change the neuron type, or mark this ensemble as off chip with `net.config[ens].on_chip = False`. From a user perspective, though, I would think "why not just detect that this can't run on Loihi and do it automatically?" So it's a balance between convenience and explicitness.

We talked in the past about having some easy way to print out a summary of what's on the chip and how it's organized. I feel like when we have that, we'll want this behavior, but perhaps until then we want to go the explicit route. But I think the explicit route includes the explicit error I mention in the previous paragraph.

Reply:
:+1: for explicit error now, and then handling it automatically once we have a better way to convey information about what is running on/off chip.

Reply:
Okay, I reverted this change and added the build error in 8f202dd64fd4b89a3ee400dc27ec8f492e3a67e7.

Reply:
I like that fix. For me, I guess I don't expect `nengo_loihi` to do a lot of moving stuff around for me to get my model to work. If I have a neuron model that doesn't work on Loihi, I'm more apt to change that neuron model than to move the ensemble off-chip. I like that this change makes the decision clear to the user, and makes them be explicit about whatever decision they make.